### PR TITLE
Feat/amm trade trait

### PR DESCRIPTION
**Pull request by drbh**
Hey guys, this PR implements the trade trait. It does not include protocol fees, since I have some questions about the `ProtocolFeeHandler`. Also I'm not sure if i've calculated LP fees correctly...

Since LP fees are calculated by 
```
let fees = T::LpFee::get() * amount_in;
```

If `amount_in` is less then `1/T::LpFee::get()`, the fee will be < 1, and is returned as 0. 

Is this the intended functionality of LP fees?

Please let me know what changes should be made

### CODE REVIEW

File: pallets/amm/src/lib.rs [hunk 1]
```diff
@@ -74,6 +80,16 @@ pub mod pallet {
...
+
+        /// Defines the fees taken out of each trade and sent back to the AMM pool,
+        /// typically 0.3%.
+        type LpFee: Get<Perbill>;
+
+        /// How much the protocol is taking out of each trade.
+        type ProtocolFee: Get<Perbill>;
+
+        // -- TODO --
+        // type ProtocolFeeHandler: OnUnbalanced<CreditOf<Self, I>>;
```
Comment (ETeissonniere) [inline pallets/amm/src/lib.rs:92]: Both ORML Tokens and Parity's Assets support the new `fungibles::*` traits from Substrate. We can use those to generate the `CreditOf` Imbalance and pass it to an `OnUnbalanced` runtime implementation

Comment (drbh) [reply pallets/amm/src/lib.rs:92]: @ETeissonniere can you help me determine what type `ProtocolFeeHandler` should be? I've tried the following: 
```rust
type ProtocolFeeHandler: OnUnbalanced<CreditOf<Self::AccountId, Self>>;
```

but get this error:
```rust
error[E0277]: the trait bound `Self: hidden_include::traits::fungibles::Unbalanced<<Self as frame_system::Config>::AccountId>` is not satisfied
```
Thank you!

Comment (ETeissonniere) [reply pallets/amm/src/lib.rs:92]: @drbh I think the last `Self` should actually be the tokens pallet

File: pallets/amm/src/lib.rs [hunk 2]
```diff
@@ -455,4 +476,119 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
             .checked_div(base_pool)
             .expect("cannot overflow with positive divisor; qed")
     }
+
+    pub fn trade(
```
Comment (ETeissonniere) [inline pallets/amm/src/lib.rs:480]: This would need to be moved to an implementation of the `AMM` trait

**Review by ETeissonniere — APPROVED** [review]

File: pallets/amm/src/lib.rs [hunk 3]
```diff
@@ -465,3 +483,127 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
...
+                // 6. Wire amount_in of the input token (identified by pair.0) from who to PalletId
+                T::Currency::transfer(
+                    pair.0,
+                    &who,
+                    &Self::account_id(),
+                    amount_without_protocol_fees,
+                )?;
+
+                // 7. Wire amount_out of the output token (identified by pair.1) to who from PalletId
+                T::Currency::transfer(pair.1, &Self::account_id(), &who, amount_out)?;
```
Comment (ashi20192020ayush) [inline pallets/amm/src/lib.rs:588]:
```suggestion
                T::Currency::transfer(quote_asset, &Self::account_id(), &who, amount_out)?;
```

File: pallets/amm/src/lib.rs [hunk 4]
```diff
@@ -465,3 +483,127 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
...
+
+                // 8. Wire protocol fees as needed (input token)
+                T::Currency::transfer(pair.0, &who, &T::ProtocolFeeReceiver::get(), protocol_fees)?;
```
Comment (ashi20192020ayush) [inline pallets/amm/src/lib.rs:591]:
```suggestion
                T::Currency::transfer(base_asset, &who, &T::ProtocolFeeReceiver::get(), protocol_fees)?;
```

File: pallets/amm/src/tests.rs [hunk 5]
```diff
@@ -366,3 +366,164 @@ fn remove_liquidity_with_more_liquidity_should_not_work() {
...
+        let trader = AccountId::from(4_u64);
+
+        // create pool and add liquidity
+        AMM::add_liquidity(
+            Origin::signed(3.into()),
+            (DOT, XDOT),
+            (100_000_000, 100_000_000),
+            (99_999, 99_999),
+        )
+        .expect("Error initalizing AMM");
```
Comment (ashi20192020ayush) [inline pallets/amm/src/tests.rs:384]: Why it would throw error?

**Review by ashi20192020ayush — CHANGES_REQUESTED** [review]

File: pallets/amm/src/tests.rs (hunk 5 above)
Comment (drbh) [reply pallets/amm/src/tests.rs:384]: it shouldn't! 🙂 I'll update them to use `asset_ok!` like the earlier examples.

File: pallets/amm/src/lib.rs (hunk 3 above)
Comment (drbh) [reply pallets/amm/src/lib.rs:588]: I don't think that `pair.0` and `quote_asset` are always the same. 

We know that `pair.0` will be the input currency from the trade, where `quote_asset` is the RHS asset in our pool. It's possible that a trader will want to input the `quote_asset` of a pool and get the `base_asset` in return. 

Due to this we can not assume that `pair.0 === quote_asset`

File: pallets/amm/src/lib.rs (hunk 4 above)
Comment (drbh) [reply pallets/amm/src/lib.rs:591]: same reasoning as the comment above.  

Earlier in the `trade` function we use the following if statement: 

```rust
// supply_in == liquidity_amount.base_amount unless inverted
let (supply_in, supply_out) = if is_inverted {
    (liquidity_amount.quote_amount, liquidity_amount.base_amount)
} else {
    (liquidity_amount.base_amount, liquidity_amount.quote_amount)
};
```

The above statement handles the case where `pair.0 != base_asset`. Therefore we'll need access the values from the tuple directly

File: pallets/amm/src/lib.rs (hunk 3 above)
Comment (ashi20192020ayush) [reply pallets/amm/src/lib.rs:589]: Can we extract variables from pair? using pair.1 and pair.2 doesn't look good. I also used in the past but team didn't like this. Please see this https://github.com/parallel-finance/parallel/pull/450#discussion_r695339486

File: pallets/amm/src/tests.rs (hunk 5 above)
Comment (ashi20192020ayush) [reply pallets/amm/src/tests.rs]: please do the same for other test-cases as well

File: pallets/amm/src/lib.rs (hunk 3 above)
Comment (drbh) [reply pallets/amm/src/lib.rs:589]: yep thats a good idea 👌

**Review by ashi20192020ayush — APPROVED** [review]

**Pull request merged**